### PR TITLE
Fix routing of events related to replies to unroutable messages.

### PR DESCRIPTION
**Pull request by hodgestar**
Errors seen that seem relevant:

```
UnroutableMessageError: ('Could not find next hop for event.', ...
```

and

```
'Error routing message for billing_dispatcher_ro'
```

and maybe

```
UnroutableMessageError: ('Outbound message for event has no associated user account:
```

### CODE REVIEW

**Comment by hodgestar** [comment]
Ready for review.

File: go/vumitools/routing.py [hunk 1]
```diff
@@ -139,6 +152,38 @@ def next_router_endpoint(self):
...
+        Returns whether the message is marked as a reply to an unroutable
+        inbound message (or an event for such a reply).
+        """
+        return self._msg['routing_metadata'].get(
+            'is_reply_to_unroutable', False)
+
+    def unroutable_event_done(self):
+        """
+        Return True if the message is an event for an unroutable reply
+        and it's hops are done.
```
Comment (jerith) [inline go/vumitools/routing.py]: Errant apostrophe.

File: go/vumitools/routing.py [hunk 2]
```diff
@@ -139,6 +152,38 @@ def next_router_endpoint(self):
...
+        """
+        if not self.get_unroutable_reply():
+            return False
+        hops = self.get_hops()
+        outbound_hops = self.get_outbound_hops()
+        if not hops and not outbound_hops:
+            return True
+        if not hops or not outbound_hops:
+            return False
```
Comment (jerith) [inline go/vumitools/routing.py:186]: I'm struggling to follow these two conditions. Can you add some explanatory comments?

Comment (hodgestar) [inline go/vumitools/routing.py:186]: These handle all the cases where `hops` or `outbound_hops` are
`None`. Will add note.

Comment (jerith) [inline go/vumitools/routing.py:186]: Also, do we differentiate between messages and events here?

File: go/vumitools/routing.py [hunk 3]
```diff
@@ -721,21 +781,32 @@ def _set_event_metadata(self, event):
...
         # set the tag on the event so that if it is from a transport
         # we can set the source of the message correctly in acquire_source.
         event_mdh.set_tag(msg_mdh.tag)
-        event_mdh.set_user_account(msg_mdh.get_account_key())
-        msg_rmeta = RoutingMetadata(msg)
-        event_rmeta.set_outbound_hops(msg_rmeta.get_hops())
+
+        if not msg_unroutable or msg_hops:
+            # unroutable replies without hops were never associated with a
+            # user account and so aren't require to have on. All other
```
Comment (jerith) [inline go/vumitools/routing.py]: "aren't required to have one."

File: go/vumitools/routing.py [hunk 4]
```diff
@@ -303,11 +348,11 @@ def get_config(self, msg):
             tag_info = yield msg_mdh.get_tag_info()
             user_account_key = tag_info.metadata['user_account']
             if user_account_key is None:
-                raise UnroutableMessageError(
+                raise UnownedTagError(
                     "Message received for unowned tag.", msg)
         else:
             raise UnroutableMessageError(
-                "Could not determine user account key", msg)
+                "No user account key or tag on message", msg)
```
Comment (jerith) [inline go/vumitools/routing.py:359]: This code path doesn't have test coverage.

File: go/vumitools/routing.py [hunk 5]
```diff
@@ -681,7 +741,7 @@ def process_outbound(self, config, msg, connector_name):
 
         target = self.find_target(config, msg, src_conn)
         if target is None:
-            raise UnroutableMessageError(
+            raise NoTargetError(
                 "No target found for outbound message from '%s': %s" % (
                     connector_name, msg), msg)
```
Comment (jerith) [inline go/vumitools/routing.py:750]: No test coverage here.

File: go/vumitools/routing.py [hunk 6]
```diff
@@ -721,21 +781,32 @@ def _set_event_metadata(self, event):
...
-        event_rmeta.set_outbound_hops(msg_rmeta.get_hops())
+
+        if not msg_unroutable or msg_hops:
+            # unroutable replies without hops were never associated with a
+            # user account and so aren't require to have on. All other
+            # messages must.
+            if not msg_mdh.has_user_account():
+                raise UnroutableMessageError(
+                    "Outbound message for event has no associated"
+                    " user account: %r" % (msg,), event)
```
Comment (jerith) [inline go/vumitools/routing.py:812]: No test coverage here.

File: go/vumitools/routing.py (hunk 4 above)
Comment (hodgestar) [inline go/vumitools/routing.py:359]: Test added.

File: go/vumitools/routing.py (hunk 5 above)
Comment (hodgestar) [inline go/vumitools/routing.py:750]: Test added.

File: go/vumitools/routing.py (hunk 6 above)
Comment (hodgestar) [inline go/vumitools/routing.py:812]: Test added.

File: go/vumitools/routing.py (hunk 2 above)
Comment (hodgestar) [inline go/vumitools/routing.py:186]: We don't explicitly check but only events have outbound hops. I can add an explicit check if you like.

**Comment by hodgestar** [comment]
Ready for re-review.

File: go/vumitools/routing.py (hunk 2 above)
Comment (jerith) [inline go/vumitools/routing.py:186]: I think this whole module needs to be refactored or rewritten for clarity -- there are way too many places where it's unclear what's happening or what kinds of things are being processed. That's out of scope for this PR, though.

Comment (hodgestar) [inline go/vumitools/routing.py:186]: I don't think the module is in terrible shape, especially given the complex
set of possibilities, but I'm all for making it clearer if someone has a
good idea how to.

**Comment by jerith** [comment]
:+1:

**Pull request merged**